docs/proposals/202001_thanos_query_health_handling.md

```diff
@@ -0,0 +1,86 @@
+2. Another option could be introduced such as `--store.hold-timeout` which would be `--store.unhealthy-timeout`'s brother and we would hold the StoreAPI nodes for `max(hold_timeout, unhealthy_timeout)`.
+3. Another option such as `--store.strict-mode` could be introduced which means that we would always retain the last information of the StoreAPI nodes of the last successful check.
+4. The StoreAPI node specification format that is used in `--store` could be extended to include another flag which would let specify the previous option per-specific node.
+
+Lets look through their pros and cons:
+
+* In general it would be nice to avoid adding new options so the first option seems the most attractive in this regard but it is also the most invasive one.
+* Second option increases the configuration complexity the most since you would have to think about the other option `--store.unhealthy-timeout` as well while setting it.
+* The last option is the most complex from code's perspective but it is also least invasive; however it has some downsides like the syntax for specifying store nodes becomes really ugly and hard to understand because we would have not only the DNS query type in there but also a special marker to enable that mode.
+
+If we were to graph these choices in terms of their incisiveness and complexity it would look something like this:
+
+```text
+Most incisive / Least Complex ------------ Least incisive / Most Complex
+#1           #2                                      #4
+           #3
+```
+
+After careful consideration and with the rationale in this proposal, we have decided to go with the third option. It should provide a sweet spot between being too invasive and providing our users the ability to fall-back to the old behavior.
+
+## Goals
+
+* Update the health check logic to properly handle the case when a node disappears in terms of a caching layer.
+
+## No Goals
+
+* Fixing the cache handling in the cases where a **new** StoreAPI gets added to the active store set.
+
+## Verification
+
+* Unit tests which would fire up a dummy StoreAPI and check different scenarios.
+* Ad-hoc testing.
+
+## Proposal
+
+* Add a new flag to Thanos Query `--store.strict-mode` which will make it always retain the last successfully retrieved information via the `Info()` gRPC method of **statically** defined nodes and thus always consider them part of the active store set.
+
+## Risk
+
+* Users might have problems removing the store nodes from the active store set since they will be there forever with this option set. However, one might argue that if the nodes go down then something like DNS service discovery needs to be used which would dynamically add and remove those nodes.
+
+## Work Plan
+
+* Implement the new flag `--store.strict-mode` in Thanos Query which will make it keep around statically defined nodes.
```

Comment:
Consider adding more info around whether we will make it default behaviour or opt-in?

Reply:
Added `It will be disabled by default to reduce surprises when upgrading.` here. I think that should be explicit enough :)